etcdctl/ctlv3/command/migrate_command.go

```diff
@@ -0,0 +1,359 @@
+	}
+	b, err := json.Marshal(n)
+	if err != nil {
+		ExitWithError(ExitError, err)
+	}
+	fmt.Fprintf(w, string(b))
+	for _, nn := range nodes {
+		max := writeKeys(w, nn)
+		if max > maxIndex {
+			maxIndex = max
+		}
+	}
+	return maxIndex
+}
+
+func readKeys(r io.Reader, be backend.Backend) error {
+	for {
+		length64, err := readInt64(r)
+		if err != nil {
+			if err == io.EOF {
+				return nil
+			}
+			return err
+		}
+
+		buf := make([]byte, int(length64))
+		if _, err := io.ReadFull(r, buf); err != nil {
+			return err
+		}
+
+		var kv mvccpb.KeyValue
+		err = proto.Unmarshal(buf, &kv)
+		if err != nil {
+			return err
+		}
+
+		writeToBackend(be, kv)
+	}
+}
+
+func readInt64(r io.Reader) (int64, error) {
+	var n int64
+	err := binary.Read(r, binary.LittleEndian, &n)
+	return n, err
+}
+
+func writeToBackend(be backend.Backend, kv mvccpb.KeyValue) {
+	ibytes := newRevBytes()
+	revToBytes(revision{main: kv.ModRevision}, ibytes)
+
+	d, err := kv.Marshal()
+	if err != nil {
+		log.Fatalf("mvcc: cannot marshal event: %v", err)
+	}
+
+	be.BatchTx().Lock()
+	be.BatchTx().UnsafePut([]byte("key"), ibytes, d)
+	be.BatchTx().Unlock()
+}
+
+const revBytesLen = 8 + 1 + 8
+
+func newRevBytes() []byte {
+	return make([]byte, revBytesLen)
+}
+
+func revToBytes(rev revision, bytes []byte) {
+	binary.BigEndian.PutUint64(bytes, uint64(rev.main))
+	bytes[8] = '_'
+	binary.BigEndian.PutUint64(bytes[9:], uint64(rev.sub))
+}
+
+func writeIndex(be backend.Backend, index uint64) {
+	bs := make([]byte, 8)
+	binary.BigEndian.PutUint64(bs, index)
+	be.BatchTx().Lock()
+	be.BatchTx().UnsafePut([]byte("meta"), []byte("consistent_index"), bs)
+	be.BatchTx().Unlock()
+}
+
+func startTransformer() (io.WriteCloser, io.ReadCloser) {
+	cmd := exec.Command(migrateTransformer)
+	cmd.Stderr = os.Stderr
+
+	writer, err := cmd.StdinPipe()
+	if err != nil {
+		ExitWithError(ExitError, err)
+	}
+
+	reader, rerr := cmd.StdoutPipe()
+	if rerr != nil {
+		ExitWithError(ExitError, rerr)
+	}
+
+	if err := cmd.Start(); err != nil {
+		ExitWithError(ExitError, err)
+	}
+
+	return writer, reader
+}
+
+func defaultTransformer() (io.WriteCloser, io.ReadCloser) {
+	// transformer decodes v2 keys from sr
+	sr, sw := io.Pipe()
+	// transformer encodes v3 keys into dw
+	dr, dw := io.Pipe()
+
+	decoder := json.NewDecoder(sr)
+
+	go func() {
+		defer func() {
+			sr.Close()
+			dw.Close()
+		}()
+
+		for {
+			if !decoder.More() {
+				break
+			}
+			node := &client.Node{}
+			if err := decoder.Decode(node); err != nil {
+				panic(err)
+			}
+			kv := transform(node)
+			if kv == nil {
+				continue
+			}
+
+			data, err := proto.Marshal(kv)
+			if err != nil {
+				panic(err)
+			}
+			buf := make([]byte, 8)
+			binary.LittleEndian.PutUint64(buf, uint64(len(data)))
+			if _, err := dw.Write(buf); err != nil {
+				panic(err)
+			}
+			if _, err := dw.Write(data); err != nil {
+				panic(err)
+			}
+		}
+	}()
+
+	return sw, dr
+}
+
+func transform(n *client.Node) *mvccpb.KeyValue {
```

Comment:
yes... this is what we do by default. just a very simple example transform func.